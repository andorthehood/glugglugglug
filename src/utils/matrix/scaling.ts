/**
 * Creates a matrix that scales along each axis independently.
 * @param sx Scale factor along the x axis.
 * @param sy Scale factor along the y axis.
 * @param sz Scale factor along the z axis.
 * @param dst Destination matrix to write into. When omitted a new matrix is created.
 * @returns A scaling matrix.
 */
export function scaling(
	sx: number,
	sy: number,
	sz: number,
	dst: number[] = new Array(16),
) {
	dst[0] = sx;
	dst[1] = 0;
	dst[2] = 0;
	dst[3] = 0;
	dst[4] = 0;
	dst[5] = sy;
	dst[6] = 0;
	dst[7] = 0;
	dst[8] = 0;
	dst[9] = 0;
	dst[10] = sz;
	dst[11] = 0;
	dst[12] = 0;
	dst[13] = 0;
	dst[14] = 0;
	dst[15] = 1;
	return dst;
}
