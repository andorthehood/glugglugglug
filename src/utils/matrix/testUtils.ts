export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
