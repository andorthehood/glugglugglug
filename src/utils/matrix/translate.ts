import { multiply } from "./multiply";
import { translation } from "./translation";

/**
 * Multiplies a matrix by a translation transform.
 * @param m Matrix to transform.
 * @param tx Translation along the x axis.
 * @param ty Translation along the y axis.
 * @param tz Translation along the z axis.
 * @param dst Destination matrix to write into. Defaults to mutating the supplied matrix.
 * @returns The transformed matrix.
 */
const translationMatrix = new Array<number>(16);

export function translate(
	m: number[],
	tx: number,
	ty: number,
	tz: number,
	dst: number[] = m,
) {
	translation(tx, ty, tz, translationMatrix);
	return multiply(m, translationMatrix, dst);
}
