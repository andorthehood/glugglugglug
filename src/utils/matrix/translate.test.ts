import { describe, expect, it } from "vitest";
import { translate } from "./translate";
import { scaling } from "./scaling";
import { formatMatrix } from "./testUtils";

describe("translate", () => {
	it("translates a matrix via the helper", () => {
		const base = scaling(1, 2, 3);

		const result = translate(base, 4, 5, 6);

		expect(result).toBe(base);
		expect(formatMatrix(result)).toMatchSnapshot();
	});

	it("writes into a provided destination", () => {
		const base = scaling(1, 2, 3);
		const dst = new Array<number>(16);
		const result = translate(base, 4, 5, 6, dst);

		expect(result).toBe(dst);
		expect(formatMatrix(dst)).toMatchSnapshot();
		expect(formatMatrix(base)).toEqual([
			[1, 0, 0, 0],
			[0, 2, 0, 0],
			[0, 0, 3, 0],
			[0, 0, 0, 1],
		]);
	});
});
