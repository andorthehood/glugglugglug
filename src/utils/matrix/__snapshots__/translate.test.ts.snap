// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translate > translates a matrix via the helper 1`] = `
[
  [
    1,
    0,
    0,
    0,
  ],
  [
    0,
    2,
    0,
    0,
  ],
  [
    0,
    0,
    3,
    0,
  ],
  [
    4,
    10,
    18,
    1,
  ],
]
`;

exports[`translate > writes into a provided destination 1`] = `
[
  [
    1,
    0,
    0,
    0,
  ],
  [
    0,
    2,
    0,
    0,
  ],
  [
    0,
    0,
    3,
    0,
  ],
  [
    4,
    10,
    18,
    1,
  ],
]
`;
